### PR TITLE
fix(#19132): fix component tree defaultExpandAll does not work when u…

### PR DESCRIPTION
**Pull request by kavience**
…sing treeData

### 🤔 This is a ...

- [ ] New feature
- [x] Bug fix
- [ ] Site / document update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |           |
| 🇨🇳 Chinese |  修复 DirectoryTree 组件, 该 bug 导致当传入属性 treeData 导致时, defaultExpandAll 不生效   |

### ☑️ Self Check before Merge

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

**Comment by afc163** [comment]
Could you fix lint and add a test case?

**Comment by kavience** [comment]
> Could you fix lint and add a test case?

ok, I will

File: components/tree/DirectoryTree.tsx [hunk 1]
```diff
@@ -82,7 +87,15 @@ class DirectoryTree extends React.Component<DirectoryTreeProps, DirectoryTreeSta
...
-      this.state.expandedKeys = getFullKeyList(props.children);
+      const expandedKeysByChildren = getFullKeyList(props.children);
+      this.state.expandedKeys = expandedKeysByChildren;
+      if (
+        !!props.treeData &&
+        Array.isArray(expandedKeysByChildren) &&
+        expandedKeysByChildren.length === 0
+      ) {
+        this.state.expandedKeys = getFullKeyListByTreeData(props.treeData);
+      }
```
Comment (zombieJ) [inline components/tree/DirectoryTree.tsx]: change to:
```jsx
if (treeData) {
  this.state.expandedKeys = ...
} else {
  this.state.expandedKeys = ...
}
```

`treeData` priority should be higher than children.

File: components/tree/util.ts [hunk 2]
```diff
@@ -101,3 +101,12 @@ export function convertDirectoryKeysToNodes(
   });
   return nodes;
 }
+
+export function getFullKeyListByTreeData(treeData: any[]): any {
+  for (let i = 0; i < treeData.length; i++) {
```
Comment (zombieJ) [inline components/tree/util.ts]: `(treeData || []).forEach` or `(treeData || []).reduce` instead.

File: components/tree/__tests__/util.test.js [hunk 3]
```diff
@@ -32,4 +32,51 @@ describe('Tree util', () => {
...
+          {
+            title: 'File',
+            key: '0-0-6',
+            isLeaf: true,
+          },
+        ],
+      },
+    ];
+
+    const keys = getFullKeyListByTreeData(treeData);
```
Comment (afc163) [inline components/tree/__tests__/util.test.js]: You should test Tree component behavior instead of testing inner function.

Comment (kavience) [reply components/tree/__tests__/util.test.js]: I am sorry, this is my first contribution for ant-design, I have see documents but I still don't understand how to write this test, can you help me ?

Comment (kavience) [reply components/tree/__tests__/util.test.js]: should I white test case in directory.test.js ? like this:
```

it('defaultExpandAll when use treeData', () => {
    const treeData = [
      {
        key: '0-0-0',
        title: 'Folder',
        children: [
          {
            title: 'Folder2',
            key: '0-0-1',
            children: [
              {
                title: 'File',
                key: '0-0-2',
                isLeaf: true,
              },
            ],
          },
        ],
      },
      {
        key: '0-0-3',
        title: 'Folder',
        children: [
          {
            title: 'File',
            key: '0-0-4',
            isLeaf: true,
          },
          {
            title: 'File',
            key: '0-0-5',
            isLeaf: true,
          },
          {
            title: 'File',
            key: '0-0-6',
            isLeaf: true,
          },
        ],
      },
    ];
    const wrapper = render(createTree({ defaultExpandAll: true, treeData }));
    expect(wrapper).toMatchSnapshot();
  });
```

Comment (afc163) [reply components/tree/__tests__/util.test.js]: Just simulate the exact problem in test case, like test case in https://github.com/ant-design/ant-design/blob/master/components/tree/__tests__/directory.test.js or https://github.com/ant-design/ant-design/blob/fb5d709ccdd8d1164b7b59982ea5ab9b9848c4c4/components/menu/__tests__/index.test.js#L327

Comment (kavience) [reply components/tree/__tests__/util.test.js]: so this is right ?  I don't sure...
```
it('DirectoryTree should expend all when use treeData and defaultExpandAll is true', () => {
    const treeData = [
      {
        key: '0-0-0',
        title: 'Folder',
        children: [
          {
            title: 'Folder2',
            key: '0-0-1',
            children: [
              {
                title: 'File',
                key: '0-0-2',
                isLeaf: true,
              },
            ],
          },
        ],
      },
    ];
    const wrapper = render(createTree({ defaultExpandAll: true, treeData }));
    expect(wrapper).toMatchSnapshot();
  });
```

Comment (afc163) [reply components/tree/__tests__/util.test.js]: If this test case can reflect the problem you meet, then it is right.

**Pull request merged**